### ISSUE TITLE
JSONP and reflected flash file (new flaw to possibly check for)

### ISSUE THREAD

**Issue by EiNSTeiN-** [presidentbeef/brakeman#474]
There's a same-origin bypass made possible by reflecting a user-supplied `param` as the first thing in a response body. More info here: http://quaxio.com/jsonp_handcrafted_flash_files/

Basically, if you return a user-supplied value as the first thing in a response body (as is usually the case for JSONP responses), the param can be made to be a valid SWF file. If this page is used in a `<object src="http://foo.com/endpoint.json?callback=...swf file...">` tag, the swf reflected by `foo.com` will have same-origin access to `foo.com`, allowing it to read csrf token, etc.

I'm not sure what would be the best way to detect this, but in my case the vulnerable app was doing something like this:
`response.body = "#{param[:callback]}"`

**Recommended fix**: the recommended fix is to (1) add a javascript comment before the reflected callback name and (2) limit the callback name length to a short size (32 characters or so). A non-vulnerable response would look like this: `response.body = "/**/#{param[:callback]}"`.

Sanitizing the parameter name to make sure it contains only alpha numerical values is _not enough_.

**Comment by jjarmoc-mtso** [comment]
A new bypass for the client-side fixes has been found as well.
http://topolik-at-work.blogspot.com/2015/06/cve-2015-3096-rosetta-flash-fix-bypass.html

**Comment by jjarmoc** [comment]
A new bypass for the client-side fixes has been found as well.
http://topolik-at-work.blogspot.com/2015/06/cve-2015-3096-rosetta-flash-fix-bypass.html

**Comment by EiNSTeiN-** [comment]
@jjarmoc this is the same flaw. I opened this issue before it was named "rosetta flash".

**Comment by jjarmoc** [comment]
I'm aware the flaw predates that name, I was commenting in regards to the discussion earlier that it's "kind of fixed client side."

**Comment by bbaja42** [comment]
Interested as well for this, though for a similar attack vector
for reflected file download attack https://drive.google.com/file/d/0B0KLoHg_gR_XQnV4RVhlNl96MHM/view

Reference: 
https://www.trustwave.com/Resources/SpiderLabs-Blog/Reflected-File-Download---A-New-Web-Attack-Vector/ 
https://www.blackhat.com/docs/eu-14/materials/eu-14-Hafif-Reflected-File-Download-A-New-Web-Attack-Vector.pdf

**Comment by presidentbeef** [comment]
RFD requires the attacker is able to set the downloaded file extension, right? I'm not sure how Brakeman would be able to detect that.

**Comment by bbaja42** [comment]
True, but reflecting any sort of input back to the user is something it might be able detect?
